Tests/Common/Orders/TimeInForces/TimeInForceTests.cs

```diff
@@ -0,0 +1,560 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using NUnit.Framework;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Orders;
+using QuantConnect.Orders.TimeInForces;
+using QuantConnect.Securities;
+using QuantConnect.Securities.Crypto;
+using QuantConnect.Securities.Equity;
+using QuantConnect.Securities.Forex;
+using QuantConnect.Tests.Common.Securities;
+
+namespace QuantConnect.Tests.Common.Orders.TimeInForces
+{
+    [TestFixture]
+    public class TimeInForceTests
+    {
+        [Test]
+        public void GtcTimeInForceCompareProperly()
+        {
+            var tif1 = new GoodTilCanceledTimeInForce();
+            var tif2 = new GoodTilCanceledTimeInForce();
+
+            Assert.IsTrue(tif1 == tif2);
+            Assert.IsFalse(tif1 != tif2);
+
+            Assert.IsTrue(tif1 == TimeInForce.GoodTilCanceled);
+            Assert.IsFalse(tif1 != TimeInForce.GoodTilCanceled);
+
+            Assert.AreEqual(tif1, tif2);
+            Assert.AreEqual(tif1.Type, tif2.Type);
+
+            Assert.AreEqual(tif1, TimeInForceType.GoodTilCanceled);
+        }
+
+        [Test]
+        public void DayTimeInForceCompareProperly()
+        {
+            var tif1 = new DayTimeInForce();
+            var tif2 = new DayTimeInForce();
+
+            Assert.IsTrue(tif1 == tif2);
+            Assert.IsFalse(tif1 != tif2);
+
+            Assert.IsTrue(tif1 == TimeInForce.Day);
+            Assert.IsFalse(tif1 != TimeInForce.Day);
+
+            Assert.AreEqual(tif1, tif2);
+            Assert.AreEqual(tif1.Type, tif2.Type);
+
+            Assert.AreEqual(tif1, TimeInForceType.Day);
+        }
+
+        [Test]
+        public void GtdTimeInForceCompareProperly()
+        {
+            var tif1 = new GoodTilDateTimeInForce(new DateTime(2018, 5, 21));
+            var tif2 = new GoodTilDateTimeInForce(new DateTime(2018, 5, 21));
+            var tif3 = new GoodTilDateTimeInForce(new DateTime(2018, 6, 6));
+
+            Assert.IsTrue(tif1 == tif2);
+            Assert.IsFalse(tif1 != tif2);
```

Comment:
I suspect this test will unexpectedly fail if changed to:
``` csharp
TimeInForce tif1 = new GoodTilDateTimeInForce(new DateTime(2018, 5, 21));
TimeInForce tif2 = new GoodTilDateTimeInForce(new DateTime(2018, 5, 21));
TimeInForce tif3 = new GoodTilDateTimeInForce(new DateTime(2018, 6, 6));

Assert.IsTrue(tif1 == tif2);
Assert.IsFalse(tif1 != tif2);
```

This is because the static equals operator does not support polymorphic/dynamic dispatch, so it will bind to the equals operator defined in `TimeInForce` (the one statically known) instead of the one defined in `GoodTilDateTimeInForce` -- see comments in `TimeInForce` regarding equality implementation.

Reply:
Aren't these asserts currently passing in the PR?
All equality methods and operators have been overridden in the `GoodTilDateTimeInForce` -- will be refactored shortly